### PR TITLE
Issue #7425: Set Leanplum user attributes asynchronously

### PR DESCRIPTION
**Pull request by hawkinsw**
It is possible to delay setting the user attributes until after initializing Leanplum. Because getting the values needed to fill out the user attributes can be slow, this will save us time on startup.

There is a risk, however. 

When calling `Leanplum.start()` with user attributes, the Leanplum session (on the server) is always has associated user attributes. When calling `Leanplum.start()` without user attributes and setting the user attributes later (using `Leanplum.setUserAttributes()`), the session on the server starts without any associated attributes. When a session's user attributes are set in this two-step dance, there is a time where sessions stored on the server will not have any associated attributes. As a result, it is possible that Leanplum analytics on the server for a particular session may not have valid attributes if the user's device loses network connectivity after the call to `Leanplum.start()` and before the call to `Leanplum.setUserAttributes()`. 

### Pull Request checklist

- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Screenshots**: This PR includes screenshots or GIFs of the changes made or an explanation of why it does not
- [ ] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features

### After merge
- [ ] **Milestone**: Make sure issues finished by this pull request are added to the [milestone](https://github.com/mozilla-mobile/fenix/milestones) of the version currently in development.

### To download an APK when reviewing a PR:
1. click on Show All Checks,
2. click Details next to "Taskcluster (pull_request)" after it appears and then finishes with a green checkmark,
3. click on the "Fenix - assemble" task, then click "Run Artifacts".
4. the APK links should be on the left side of the screen, named for each CPU architecture

### CODE REVIEW

File: app/src/main/java/org/mozilla/fenix/components/metrics/LeanplumMetricsService.kt [hunk 1]
```diff
@@ -70,17 +73,30 @@ class LeanplumMetricsService(private val application: Application) : MetricsServ
...
-        ))
+        Leanplum.start(application)
+
+        MainScope().launch(Dispatchers.IO) {
+            val installedApps = MozillaProductDetector.getInstalledMozillaProducts(application)
+
+            Leanplum.setUserAttributes(
+                hashMapOf(
+                    "default_browser" to
+                            MozillaProductDetector.getMozillaBrowserDefault(application).orEmpty(),
```
Comment (mcomella) [inline app/src/main/java/org/mozilla/fenix/components/metrics/LeanplumMetricsService.kt]: This PR doesn't remove the redundant computation in each of the `MozillaProductDetector` method calls, which are wasting the users battery & clogging up our background threads.

**Comment by hawkinsw** [comment]
> Curious to ask since I can't find an answer after extensively googling... Looking at the Coroutine Docs, and seeing how the `MainScope()` [is used](https://github.com/Kotlin/kotlinx.coroutines/blob/master/docs/coroutine-context-and-dispatchers.md#coroutine-scope), would we have to explicitly cancel it in this case?

Very good point. We discussed this asynchronously on slack. Thanks for bringing it up.

**Comment by mcomella** [comment]
I have a few thoughts.

Are we sure this is safe to call when not on the UI thread? I noticed that `Leanplum.start` is not documented as thread safe. Looking at the code, I see it assigning to a few variables that are not synchronized, nor `@Volatile`.

Are we seeing worthwhile gains? In a local profile before this patch, inside `LeanplumMetricsService.start`, I see:
* This is the first method to prime the `BrowsersCache` (20.75ms on Pixel 2)
* `TuplesKt.to` is visible but takes a minimal amount of time (1.23 ms on Pixel 2)

I don't think this is significant enough to put on a background thread but maybe I missed something. However, if we wanted to improve this, I think it might be safer to "prime" `BrowsersCache` on a background thread in anticipation of this code, due to the reasons above and below.

---

Thanks for the explanation of the possible side effects, hawkinsw. I noticed an additional one [in the docs](https://docs.leanplum.com/reference#user-attributes):

> // Passing attributes at session start allows us to target content based on the attributes.
> Leanplum.start(userAttributes: ["gender":"Female", "age": 29])
>
> // You can also pass them later on in the session, but you won't be able to
> // target variables or messages at these for that session.
> Leanplum.setUserAttributes(["gender":"Female", "age": 29])

Earlier on in the docs, they mention:

> Each session has its own user attributes, but they get copied from one session to the next.

And [elsewhere](https://docs.leanplum.com/reference#sessions) they mention that sessions can be paused/resumed and how new sessions are created:

> When the app is backgrounded, Leanplum will pause the session. When the app is re-opened, the session will resume. If the pause (background) lasts over 30 minutes, resuming the session will count as a new session. 

It sounds like our first session would not be targetable on these attributes and a new session will not create until 1) (maybe, based on when `Leanplum.start` is called) the process is restarted or 2) the app hasn't been used in 30 minutes.

I'm concerned this might not meet our product requirements. That being said, if we can resolve the other concerns, we can bring this to our Leanplum consumers and discuss the trade-offs.

---

@hawkinsw What are your thoughts?

**Comment by hawkinsw** [comment]
> I have a few thoughts.
> 
> Are we sure this is safe to call when not on the UI thread? I noticed that `Leanplum.start` is not documented as thread safe. Looking at the code, I see it assigning to a few variables that are not synchronized, nor `@Volatile`.

I don't see the problem. We are not calling `.start()` off MT with this patch. I must not be understanding correctly what you are saying. 

Calling `.start()` off MT is *definitely* not safe. We have documented the issues with doing *that* in other issues. However, doing `.setUserAttributes()` off MT is okay, I think.

> 
> Are we seeing worthwhile gains? In a local profile before this patch, inside `LeanplumMetricsService.start`, I see:
> 
>     * This is the first method to prime the `BrowsersCache` (20.75ms on Pixel 2)
> 
>     * `TuplesKt.to` is visible but takes a minimal amount of time (1.23 ms on Pixel 2)
> 
> 
> I don't think this is significant enough to put on a background thread but maybe I missed something. However, if we wanted to improve this, I think it might be safer to "prime" `BrowsersCache` on a background thread in anticipation of this code, due to the reasons above and below.
> 
> Thanks for the explanation of the possible side effects, hawkinsw. I noticed an additional one [in the docs](https://docs.leanplum.com/reference#user-attributes):
> 
> > // Passing attributes at session start allows us to target content based on the attributes.
> > Leanplum.start(userAttributes: ["gender":"Female", "age": 29])
> > // You can also pass them later on in the session, but you won't be able to
> > // target variables or messages at these for that session.
> > Leanplum.setUserAttributes(["gender":"Female", "age": 29])
> 
> Earlier on in the docs, they mention:
> 
> > Each session has its own user attributes, but they get copied from one session to the next.
> 
> And [elsewhere](https://docs.leanplum.com/reference#sessions) they mention that sessions can be paused/resumed and how new sessions are created:
> 
> > When the app is backgrounded, Leanplum will pause the session. When the app is re-opened, the session will resume. If the pause (background) lasts over 30 minutes, resuming the session will count as a new session.
> 
> It sounds like our first session would not be targetable on these attributes and a new session will not create until 1) (maybe, based on when `Leanplum.start` is called) the process is restarted or 2) the app hasn't been used in 30 minutes.
> 
> I'm concerned this might not meet our product requirements. That being said, if we can resolve the other concerns, we can bring this to our Leanplum consumers and discuss the trade-offs.
> 
> @hawkinsw What are your thoughts?

This, more operational issue, seems like the real question. Great catch! Should we bring in someone from the team that actually uses Leanplum to decide whether or not this is necessary?

**Comment by mcomella** [comment]
Next steps: 
- @mcomella will check with Leanplum consumers if this trade-off is acceptable (yes/no answer, they shouldn't need to think about it)
- If so, we should investigate if the change is worth making
- If it is, we can investigate the viability of making the change further

**Comment by mcomella** [comment]
@hawkinsw Our Leanplum consumers mentioned we need _some_ user attributes for the first session. In particular:

> important ones: default browser, fennec installed, search widget installed

Given the increasing scope and my earlier Pixel 2 measurement where I didn't notice any significant wins, I'm inclined to say we defer this in favor of higher impact work but I'm happy to hear your thoughts.

**Comment by hawkinsw** [comment]
Agree 100 percent. Thank you for confirming with the LP consumers.

**Comment by mcomella** [comment]
I'm closing to close this and move the issue back to "Needs triage". We can reopen if we revisit the issue. Thanks for putting this together, hawkinsw.

**Comment by mcomella** [comment]
This PR pointed at #7425, which was fixed by the `BrowsersCache` change, so I filed a new issue: https://github.com/mozilla-mobile/fenix/issues/7704

**Pull request closed without merging**